### DISCUSSION TITLE
Mailto links in markdown

### DISCUSSION THREAD

**Discussion by TheBrambleShark** [statiqdev/Discussions#160] [Q&A]
Hello,

Hopefully this is a simple question and just something I missed. I have a markdown link that looks like this
```md
[trademark@luzfaltex.com](mailto:trademark@luzfaltex.com)
```

When rendered in the webpage, it gets rendered as so:
```html
<a href="mailto:trademark&#64;luzfaltex.com">trademark@luzfaltex.com</a>
```
I would expect that text inside of a link would not be touched, but it appears this is not the case. I have tried escaping the `@` with `\@` but this had no effect on the issue.

Any ideas how I can fix this?

**Comment by girlpunk** [comment]
Did you manage to find a solution to this?

**Reply by TheBrambleShark** [reply]
Unfortunately I have not. I also haven't touched this project in a while, so I may try revisiting it to find a workaround.

**Comment by daveaglick** [comment]
Sorry for the long delay - I'm trying to get back on track open source wise, but this question fell through the cracks. I'm going to copy my answer from https://github.com/statiqdev/Statiq.Framework/issues/254 wholesale here since it's valuable information for the discussion board...

---

There's a tricky dance that happens with escaping the `@` symbol and Razor parsing. Even though escaping `@` isn't necessary for plain old Markdown, it's so common to run Statiq documents through the Razor processor, that escaping them after Markdown rendering is the default behavior.

If you're creating your own pipelines and want to turn off automatic `@` escaping entirely (which is only advisable when _not_ using Razor downstream of the Markdown module), you can call `EscapeAt(false)` on the `RenderMarkdown` module:

```
// ...
new RenderMarkdown().EscapeAt(false);
// ...
```

Otherwise there's really no way to tell when an `@` symbol in post-Markdown content is intended to be a Razor instruction or was from the Markdown output, so they all get escaped. If you want to disable this behavior for a specific `@` symbol as with a `mailto:` link, you can escape it with `\@`:

```
[test\@examples.com](mailto:test\@example.com)
```

That produces:

```
<a href="mailto:test&#64;example.com">test@examples.com</a>
```

Which obviously isn't _quite_ right because the `@` symbol in the `a` element is still escaped. That's because of the way the Markdown renderer sequences the generated link element output vs. the content rendering, and where the escaping happens.

To get _exactly_ what you want, you can put the mail link HTML right in the Markdown document, with the `\@` escaping:

```
<a href="mailto:test\@example.com">test\@example.com</a>
```

Which produces the desired unescaped output:

```
<a href="mailto:test@example.com">test@example.com</a>
```

Note that this all appears to be academic. The client should correctly interpret the HTML entities and convert them back to the correct characters:

![image](https://user-images.githubusercontent.com/1020407/206793890-d106ae24-69d9-49fb-82d5-581ecdd5a841.png)

Is that not what you're seeing? I.e. in the rendered output does the `mailto:` link with `#64;` entities not work when clicked or display to the user incorrectly?

**Reply by girlpunk** [reply]
> Is that not what you're seeing? I.e. in the rendered output does the mailto: link with https://github.com/orgs/statiqdev/discussions/64; entities not work when clicked or display to the user incorrectly?

Unfortunately not, the unescaped value (i.e. test&#64;example.com) shows in the link preview, and in the To box of the new email opened upon clicking the link.

**Reply by daveaglick** [reply]
I think I might be missing something...isn't the unescaped value (I.e. with an `@` instead of the HTML entity code `#64;`) what you want to be seeing in the link preview and To box?

**Reply by girlpunk** [reply]
Sorry, I think I made a typo (unescaped should be escaped), and I guess github sometimes parses those entity codes? because I'm fairly sure I typed `test&#64;example.com` (
![image](https://user-images.githubusercontent.com/161093/207386550-ba3f380b-bef7-4625-811f-dc619b9c5595.png)
)

**Reply by daveaglick** [reply]
Haha, foiled by GitHub!

So that's interesting then - I would have expected the behavior when rendering and presenting escaped HTML entities to behave the same across browsers, though maybe it's one of those weird poorly-defined parts of the standard that results in browser differences. What browser are you using? Can you see if a different browser yields the same results for you? I'll do the same here with the browser you use and see if I can reproduce that behavior.

**Reply by girlpunk** [reply]
I'm currently using Edge 108.0.1462.46. Will have to email security about access to other browsers but I'll let you know.

**Reply by daveaglick** [reply]
Huh, now I'm totally confused - the screen shot above was taken from Edge. And here's another I checked just now to verify something funny wasn't happening over the wire to replace the HTML entity escapes:

![image](https://user-images.githubusercontent.com/1020407/207391327-4352a542-1cf2-4586-95a1-1cc5c0e5465f.png)

I checked on a few other browsers and two completely different systems I have access to, all the same.

I've been assuming you were on Windows (especially using Edge), but is that wrong - are you on a Mac? Not that it would do anything different HTML-rendering wise, but I'm at a total loss at this point why your browser wouldn't be correctly interpreting those HTML entities.

**Reply by girlpunk** [reply]
As am I, because that's different from what I'm seeing:
![image](https://user-images.githubusercontent.com/161093/209134341-45aab27d-6326-4070-ae60-5885e0b8f206.png)

You assumed correctly, Edge Version 108.0.1462.54 (Official build) (64-bit) running on Windows 10 Version 22H2 (OS Build 19045.2364)

**Reply by girlpunk** [reply]
Ah, might have something. I've just spotted that your screenshot was from the network tab whereas mine is the elements. Looking at the same location in the network response shows this:

```html
<a href="mailto:support&amp;#64;sxxxx.com">support&#64;sxxxx.com</a>
```

Note the double escaping on the entity in the link reference
